packages/email-petition/src/city-autocomplete.tsx

```diff
@@ -0,0 +1,172 @@
+"use client";
+
+import { useState, useEffect, useRef } from "react";
+import { Input } from "@dxe/petitions-components/input";
+
+interface CityPrediction {
+  cityAddress: string;
+  cityName: string;
+}
+
+interface CityAutocompleteProps {
+  value: string;
+  onChange: (value: string) => void;
+  onBlur: () => void;
+  disabled?: boolean;
+  placeholder?: string;
+  lat?: number;
+  lng?: number;
+}
+
+export function CityAutocomplete({
+  value,
+  onChange,
+  onBlur,
+  disabled = false,
+  placeholder = "Santa Rosa",
+  lat,
+  lng,
+}: CityAutocompleteProps) {
+  const [predictions, setPredictions] = useState<CityPrediction[]>([]);
+  const [isOpen, setIsOpen] = useState(false);
+  const inputRef = useRef<HTMLInputElement>(null);
+  const dropdownRef = useRef<HTMLDivElement>(null);
+
+  const fetchPredictions = async (input: string) => {
+    if (input.length == 0 || !lat || !lng) {
+      setPredictions([]);
+      setIsOpen(false);
+      return;
+    }
+
+    try {
+      const apiKey = process.env.NEXT_PUBLIC_GOOGLE_MAPS_PLACES_NEW_API_KEY;
+      if (!apiKey) {
+        console.error("Google Maps API key not found");
+        return;
+      }
+
+      // Generate session token for cost optimization (max 36 characters)
+      const sessionToken = `ac-${Date.now().toString(36)}-${Math.random().toString(36).substring(2, 5)}`;
+
+      // Request body for Google Places API
+      const requestBody = {
+        input,
+        locationRestriction: {
+          circle: {
+            center: { latitude: lat, longitude: lng },
+            radius: 50000,
+          },
+        },
+        includedPrimaryTypes: ["locality"],
+        sessionToken,
+      };
+
+      const response = await fetch(
+        `https://places.googleapis.com/v1/places:autocomplete?key=${apiKey}`,
+        {
+          method: "POST",
+          headers: {
+            "Content-Type": "application/json",
+            "X-Goog-FieldMask":
+              "suggestions.placePrediction.text,suggestions.placePrediction.structuredFormat.mainText.text",
+          },
+          body: JSON.stringify(requestBody),
+        },
+      );
+
+      if (!response.ok) {
+        throw new Error(`HTTP error! status: ${response.status}`);
+      }
+
+      const data = await response.json();

```

Comment:
please extract into its own function

Reply:
oh sorry i thought i had selected a range of lines. i think it should start with (and include) the URL construction. then it's a clean boundary what the function does: it's a wrapper around the API method. naturally it takes care of HTTP / transport related details, which includes both URI construction and JSON decoding.